url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/ColliderImpl.java

```diff
@@ -0,0 +1,798 @@
+                        m_selector.selectNow();
+                    }
+                    else
+                        m_selector.select();
+                }
+                else
+                {
+                    m_selector.selectNow();
+                    if (m_selector.keys().size() == 0)
+                    {
+                        assert( readers == 0 );
+                        break;
+                    }
+                }
+
+                if (s_strTailUpdater.compareAndSet(this, null, dummyRunnable))
+                    s_strHeadUpdater.lazySet( this, dummyRunnable );
+
+                final Set<SelectionKey> selectedKeys = m_selector.selectedKeys();
+                for (SelectionKey key : selectedKeys)
+                {
+                    final ChannelHandler channelHandler = (ChannelHandler) key.attachment();
+                    readers += channelHandler.handleReadyOps( m_threadPool );
+                }
+                selectedKeys.clear();
+
+                SelectorThreadRunnable runnable;
+                while ((runnable = m_strHead) == null);
+
+                for (;;)
+                {
+                    SelectorThreadRunnable next = runnable.nextSelectorThreadRunnable;
+                    if (next == null)
+                    {
+                        m_strHead = null;
+                        if (!s_strTailUpdater.compareAndSet(this, runnable, null))
+                        {
+                            while ((next = runnable.nextSelectorThreadRunnable) == null);
+                            s_nextSelectorThreadRunnableUpdater.lazySet( runnable, null );
+                        }
+                    }
+                    else
+                        s_nextSelectorThreadRunnableUpdater.lazySet( runnable, null );
+
+                    readers -= runnable.runInSelectorThread();
+                    assert( readers >= 0 );
+
+                    runnable = next;
+                    if (runnable == null)
+                    {
+                        runnable = m_strHead;
+                        if (runnable == null)
+                            break;
+                    }
+                }
+
+                SelectorThreadRunnable strLater = m_strLater;
+                m_strLater = null;
+                while (strLater != null)
+                {
+                    runnable = strLater;
+                    strLater = runnable.nextSelectorThreadRunnable;
+                    runnable.nextSelectorThreadRunnable = null;
+                    final int rc = runnable.runInSelectorThread();
+                    assert( rc == 0 );
+                }
+
+                /* End of select loop */
+            }
+
+            m_threadPool.stopAndWait();
+        }
+        catch (final IOException ex)
+        {
+            if (s_logger.isLoggable(Level.WARNING))
+                s_logger.warning( ex.toString() );
+        }
+        catch (final InterruptedException ex)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)